adr/00005-user-authorisation-using-pundit.md

```diff
@@ -0,0 +1,84 @@
+# 5. User Authorisation using Pundit
+
+Date: 2024-02-05
+
+## Status
+
+Accepted
+
+## Context
+
+We want to have scalable user authorisation and permissions. We started with conditional scoping and per-case permissions, such as the following code snippets:
+
+- Permissions
+
+  ```erb
+  <% if user != current_user %>
+    <%= link_to "Remove", remove_user_path(user) %>
+  <% end %>
+  ```
+
+- Scopes
+
+  ```ruby
+  class SchoolsController
+    def index
+      @schools = scoped_schools
+    end
+
+    private
+
+    def scoped_schools
+      return School.all if current_user.support_user?
+
+      current_user.schools
+    end
+  end
+  ```
+
+This logic is contained within Controllers and Views. This makes the logic both difficult to unit test and re-use.
+
+## Decision
+
+- Add the [Pundit](https://github.com/varvet/pundit) gem.
+- Adopt the resource policies pattern.
+
+## Consequences
+
+Moving forward, user permissions should be defined within policy classes.
+
+The above can now be re-written as:
+
+```ruby
+class UserPolicy < ApplicationPolicy
+  def destroy?
+    current.support_user? && user != current_user
+  end
+end
+
+<% if policy(user).destroy? %>
+  <%= link_to "Remove", remove_user_path(user) %>
+<% end %>
+```
+
+and
+
+```ruby
+class SchoolPolicy < ApplicationPolicy
+  class Scope < ApplicationPolicy::Scope
+    def resolve
+      if user.support_user?
+        scope.all
+      else
+        scope.joins(:users).where(users: { id: user })
+      end
+    end
+  end
+end
+
+class SchoolsController
+  def index
+    @schools = policy_scope(School.all)
```

Comment:
I realise that `School.all.all` is a perfectly valid thing to do, but it feels odd. The [Pundit docs](https://github.com/varvet/pundit/blob/9dd051102bbce8899a2f5b059cd9692052c3bf8b/README.md?plain=1#L302) seem to prefer passing in just the top-level model in these cases.

```suggestion
    @schools = policy_scope(School)
```